### PR TITLE
Synchronizer logic

### PR DESCRIPTION
**Pull request by frankobe**

### CODE REVIEW

File: src/sync/index.js [hunk 1]
```diff
@@ -0,0 +1,344 @@
+const resolvers = require('../schema/resolvers');
+const _ = require('lodash');
+const connectDB = require('../db')
+
+const Qweb3 = require('../qweb3.js/src/qweb3');
+const qclient = new Qweb3('http://bodhi:bodhi@localhost:13889');
+
+
```
Comment (dwalintukan) [inline src/sync/index.js]: 👾

File: src/sync/index.js [hunk 2]
```diff
@@ -0,0 +1,344 @@
...
+const Vote = require('./models/vote');
+const OracleResultSet = require('./models/oracleResultSet');
+const FinalResultSet = require('./models/finalResultSet');
+
+const Contracts = require('./contracts');
+const contractEventFactory = qclient.Contract(Contracts.EventFactory.address, Contracts.EventFactory.abi);
+const contractOracleFactory = qclient.Contract(Contracts.OracleFactory.address, Contracts.OracleFactory.abi);
+const constractTopicEvent = qclient.Contract(null, Contracts.TopicEvent.abi);
+const constractCentralizedOracle = qclient.Contract(null, Contracts.CentralizedOracle.abi);
+const constractDecentralizedOracle = qclient.Contract(null, Contracts.DecentralizedOracle.abi);
```
Comment (dwalintukan) [inline src/sync/index.js]: contract, not constract

File: src/sync/models/centralizedOracle.js [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+      optionIdxs: Array.from(Array(this.numOfResults).keys()),
+      amounts: _.fill(Array(this.numOfResults), 0),
+      resultIdx: null,
+      blockNum: this.blockNum,
+      endBlock: this.bettingEndBlock
+    }
+  }
+}
+
+module.exports = CentralizedOracle;
```
Comment (dwalintukan) [inline src/sync/models/centralizedOracle.js]: extra line

File: src/sync/models/decentralizedOracle.js [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+    let intermedia = _.map(this.rawLog['_resultNames'], (item) => _.trimEnd(utils.toAscii(item), '\u0000'));
+    this.resultNames = _.filter(intermedia, item => !!item);
+
+    this.contractAddress = this.rawLog['_contractAddress'];
+    this.eventAddress = this.rawLog['_eventAddress'];
+    this.numOfResults = this.rawLog['_numOfResults'].toNumber();
+    this.lastResultIndex = this.rawLog['_lastResultIndex'].toNumber();
+    this.arbitrationEndBlock = this.rawLog['_arbitrationEndBlock'].toNumber();
+    this.consensusThreshold = this.rawLog['_consensusThreshold'].toNumber();
+
```
Comment (dwalintukan) [inline src/sync/models/decentralizedOracle.js]: 👾

File: src/sync/models/decentralizedOracle.js [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+      optionIdxs: optionIdxs,
+      amounts: _.fill(Array(this.numOfResults-1), 0),
+      resultIdx: null,
+      blockNum: this.blockNum,
+      endBlock: this.arbitrationEndBlock
+    }
+  }
+}
+
+module.exports = DecentralizedOracle;
```
Comment (dwalintukan) [inline src/sync/models/decentralizedOracle.js]: extra line

File: src/sync/models/finalResultSet.js [hunk 6]
```diff
@@ -0,0 +1,27 @@
...
+
+  translate() {
+    return {
+      topicAddress: this.eventAddress,
+      resultIdx: this.finalResultIndex,
+    }
+  }
+}
+
+module.exports = FinalResultSet;
```
Comment (dwalintukan) [inline src/sync/models/finalResultSet.js]: extra line

File: src/sync/models/oracleResultSet.js [hunk 7]
```diff
@@ -0,0 +1,27 @@
...
+
+  translate() {
+    return {
+      oracleAddress: this.oracleAddress,
+      resultIdx: this.resultIndex
+    }
+  }
+}
+
+module.exports = OracleResultSet;
```
Comment (dwalintukan) [inline src/sync/models/oracleResultSet.js]: extra line

File: src/sync/models/topic.js [hunk 8]
```diff
@@ -0,0 +1,49 @@
...
+    this.name = _.trimEnd(utils.toAscii(nameHex), '\u0000');
+    let intermedia = _.map(this.rawLog['_resultNames'], (item) => _.trimEnd(utils.toAscii(item), '\u0000'));
+    this.resultNames = _.filter(intermedia, item => !!item);
+
+    this.topicAddress = this.rawLog['_topicAddress'];
+    this.creator = this.rawLog['_creator'];
+    this.oracle = this.rawLog['_oracle'];
+    this.bettingEndBlock = this.rawLog['_bettingEndBlock'].toNumber();
+    this.resultSettingEndBlock = this.rawLog['_resultSettingEndBlock'].toNumber();
+
```
Comment (dwalintukan) [inline src/sync/models/topic.js]: 👾

File: src/sync/models/topic.js [hunk 9]
```diff
@@ -0,0 +1,49 @@
...
+      options: this.resultNames,
+      resultIdx: null,
+      qtumAmount: _.fill(Array(this.resultNames.length), 0),
+      botAmount:_.fill(Array(this.resultNames.length), 0),
+      blockNum: this.blockNum,
+    }
+  }
+}
+
+module.exports = Topic;
```
Comment (dwalintukan) [inline src/sync/models/topic.js]: extra line

File: src/sync/models/vote.js [hunk 10]
```diff
@@ -0,0 +1,32 @@
...
+      voterAddress: this.participant,
+      oracleAddress: this.oracleAddress,
+      optionIdx: this.resultIndex,
+      amount: this.votedAmount,
+      blockNum: this.blockNum,
+    }
+  }
+}
+
+module.exports = Vote;
```
Comment (dwalintukan) [inline src/sync/models/vote.js]: extra line

File: src/sync/index.js [hunk 11]
```diff
@@ -0,0 +1,344 @@
...
+
+const batch_size=100;
```
Comment (dwalintukan) [inline src/sync/index.js]: use camel case for consistency

File: src/sync/index.js [hunk 12]
```diff
@@ -0,0 +1,344 @@
...
+
+const contract_deployed_block_num = 48000;
```
Comment (dwalintukan) [inline src/sync/index.js]: use camel case for consistency

File: src/sync/index.js [hunk 13]
```diff
@@ -0,0 +1,344 @@
...
+                      db.Blocks.insert({'blockNum': i});
+                    }
+                    start_block = end_block+1;
+                    loop.next();
+                  }
+                },
+              (err) => {
+                console.log(err.message);
+                syncTopic = true;
+                if (syncTopic && syncCOracle && syncDOracle && syncVote && syncOracleResult && syncFinalResult){
```
Comment (dwalintukan) [inline src/sync/index.js]: is this necessary because all the searchlogs are asynchronous?

Comment (chrisli30) [reply src/sync/index.js]: +1

Comment (frankobe) [reply src/sync/index.js]: yes. Note that err neednt due to code error. A parsing error caused by legacy events cant be ignored. So I still mark this as true and keep looping

File: src/sync/models/centralizedOracle.js (hunk 3 above)
Comment (frankobe) [reply src/sync/models/centralizedOracle.js]: done

File: src/sync/models/decentralizedOracle.js (hunk 5 above)
Comment (frankobe) [reply src/sync/models/decentralizedOracle.js]: done

File: src/sync/models/finalResultSet.js (hunk 6 above)
Comment (frankobe) [reply src/sync/models/finalResultSet.js]: done

File: src/sync/models/oracleResultSet.js (hunk 7 above)
Comment (frankobe) [reply src/sync/models/oracleResultSet.js]: done

File: src/sync/models/topic.js (hunk 9 above)
Comment (frankobe) [reply src/sync/models/topic.js]: done

File: src/sync/models/vote.js (hunk 10 above)
Comment (frankobe) [reply src/sync/models/vote.js]: done

**Pull request merged**